src/freertos_drivers/esp32/Esp32WiFiManager.hxx

```diff
@@ -0,0 +1,316 @@
+static constexpr const char *WIFI_POWER_SAVE_DESC = 
+    "When enabled this allows the ESP32 WiFi radio to use power savings mode "
+    "which puts the radio to sleep except to receive beacon updates from the "
+    "connected SSID. This should generally not need to be enabled unless you "
+    "are powering the ESP32 from a battery.";
+
+/// Visible name for the Hub Configuration group.
+static constexpr const char *HUB_NAME = "Hub Configuration";
+
+/// Visible description for the Hub Configuration group.
+static constexpr const char *HUB_DESC = 
+    "Configuration settings for an OpenLCB Hub";
+
+/// Visible name for the hub_mode field.
+static constexpr const char *HUB_MODE_NAME = "Enable Hub Mode";
+
+/// Visible description for the hub_mode field.
+static constexpr const char *HUB_MODE_DESC = 
+    "Defines this node as a hub which can accept connections";
+
+/// Visible name for the hub_listener_port field.
+static constexpr const char *HUB_LISTENER_PORT_NAME =
+    "Hub Listener Port";
+
+/// Visible name for the hub_listener_port field.
+static constexpr const char *HUB_LISTENER_PORT_DESC = 
+    "Defines the TCP/IP listener port this node will use when operating as a "
+    "hub. Most of the time this does not need to be changed.";
+
+/// Visible name for the link_config group.
+static constexpr const char *LINK_NAME = "Node Connect Configuration";
+
+/// Visible name for the link_config group.
+static constexpr const char *LINK_DESC =
+    "Configures how this node will connect to other nodes.";
+
+CDI_GROUP(HubConfiguration);
+CDI_GROUP_ENTRY(enabled, Uint8ConfigEntry,
+    Name(HUB_MODE_NAME), Description(HUB_MODE_DESC), Min(0), Max(1),
+    Default(0), MapValues(BOOLEAN_MAP));
+CDI_GROUP_ENTRY(port, Uint16ConfigEntry,
+    Name(HUB_LISTENER_PORT_NAME),
+    Description(HUB_LISTENER_PORT_DESC),
+    Min(1), Max(65535), Default(TcpClientDefaultParams::DEFAULT_PORT))
+CDI_GROUP_ENTRY(service_name, StringConfigEntry<48>,
+    Name(TcpClientDefaultParams::SERVICE_NAME),
+    Description(TcpClientDefaultParams::SERVICE_DESCR));
+CDI_GROUP_END();
+
+CDI_GROUP(WiFiConfiguration);
+CDI_GROUP_ENTRY(wifi_sleep, Uint8ConfigEntry,
+    Name(WIFI_POWER_SAVE_NAME), Description(WIFI_POWER_SAVE_DESC),
+    Min(0), Max(1), Default(0), MapValues(BOOLEAN_MAP));
+CDI_GROUP_ENTRY(hub_config, HubConfiguration,
+    Name(HUB_NAME),
+    Description(HUB_DESC));
+CDI_GROUP_ENTRY(link_config, TcpClientConfig<TcpClientDefaultParams>,
+    Name(LINK_NAME),
+    Description(LINK_DESC));
+CDI_GROUP_END();
+
+class Esp32WiFiManager : public ConfigUpdateListener
+{
+public:
+    /// Constructor.
+    ///
+    /// @param ssid is the WiFi AP to connect to.
+    /// @param password is the password for the WiFi AP being connected to.
+    /// @param hostname_prefix is the beginning part of the hostname that will
+    /// be advertised via mDNS and to the WiFi AP.
+    /// @param node_id is this node's unique ID, the last 32 bits are appended
```

Comment:
I had that initially but with it I was seeing heap explosions (350kb free to zero). Also the id is only used here to generate the hostname. Perhaps a second constructor taking only the OpenMRN object and remove the null check paradigm since it may confuse the consumer?

Reply:
I'm going to switch to two constructors to clearly delineate functionality. But I'd like to revisit this node-id passing and drop it if possible.